### PR TITLE
Feature: Token/Fiat Input components

### PR DESCRIPTION
**Pull request by mdjastrzebski**
Scope:
- [x] New TokenValueInput, FiatValueInput and related controls (implementing new designs)
- [x] Replaced all `ValueInput` use cases (Send, Send Synth, Exchange, Key/Archanova Migrator), except legacy Liquidity Pool screens

Details:
- [x] `TextValueInput` control for entering token value, build on `BigNumberInput` (+ 2 common accessory views)
- [x] `FiatValueInput` control for entering fiat value, build on `BigNumberInput`
- [x] `BigNumberInput` control for entering valid `BigNumber` value, based on `AutoScaleTextInput`
- [x] `AutoScaleTextInput` control that scales text down (until certain point) in order to display whole text at once, based on `TextInput`
- [x] `TextInput` basic text input, with simple styling 
- [x] `TokenFiatValueInputs` a complex control for entering either token or fiat value for transfer.

### CODE REVIEW

File: src/components/Icons/TokenIcon.js [hunk 1]
```diff
@@ -67,11 +68,9 @@ function TokenIcon({ url, size = 48, chain, style, imageStyle, chainIconStyle }:
       <Image source={source} style={[imageSizeStyle, imageStyle]} />
 
       {!!ChainIcon && (
-        <ChainIcon
-          width={chainIconSize}
-          height={chainIconSize}
-          style={[styles.chainIcon, chainIconSizeStyle, chainIconStyle]}
-        />
+        <View style={[styles.chainIconWrapper, chainIconSizeStyle, chainIconStyle]}>
```
Comment (poocart) [inline src/components/Icons/TokenIcon.js:71]: wdyt if we can use `chainIconWrapper` values as styled component? I mean styles are OK, but it's not widely used throughout our codebase, that's the only take

File: src/components/Inputs/TokenValueInput/TokenBalanceAccessory.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+}: Props) => {
+  const rates = useChainRates(chain);
+  const currency = useFiatCurrency();
+
+  const handleUseMax = () => {
+    Keyboard.dismiss();
+    onUseMax?.();
+  };
+
+  if (!balance) return null;
```
Comment (poocart) [inline src/components/Inputs/TokenValueInput/TokenBalanceAccessory.js:75]: I'm thinking this vs 0 balance display?

File: src/components/SendAsset/SendContainer.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+import ContactSelector from 'components/ContactSelector';
+import Spinner from 'components/Spinner';
+
+// Constants
+import { ASSET_TYPES } from 'constants/assetsConstants';
+
+// Utils
+import { spacing } from 'utils/variables';
+import { CHAIN } from 'constants/chainConstants';
+import { useChainConfig } from 'utils/uiConfig';
```
Comment (poocart) [inline src/components/SendAsset/SendContainer.js:41]:
```suggestion
// Constants
import { ASSET_TYPES } from 'constants/assetsConstants';
import { CHAIN } from 'constants/chainConstants';

// Utils
import { spacing } from 'utils/variables';
import { useChainConfig } from 'utils/uiConfig';
```

File: src/screens/Exchange/Exchange/Exchange.js [hunk 4]
```diff
@@ -166,32 +106,46 @@ function Exchange() {
...
     return () => {
       isCancelled = true;
     };
-  }, [fromAsset, dispatch]);
+  }, [dispatch, fromAsset]);
+
+  React.useEffect(() => {
+    if (isLogV2AppEvents()) {
+      dispatch(logEventAction('v2_exchange_pair_selected'));
+    }
```
Comment (poocart) [inline src/screens/Exchange/Exchange/Exchange.js:117]: I would add `fromAsset` and `toAsset` and remove `chain`.

File: src/screens/Exchange/Exchange/Exchange.js [hunk 5]
```diff
@@ -166,32 +106,46 @@ function Exchange() {
...
+  }, [dispatch, chain, fromAddress, toAddress]);
```
Comment (poocart) [inline src/screens/Exchange/Exchange/Exchange.js:118]: I would replace addresses with fromAsset and toAsset and remove chain.

File: src/components/Inputs/TokenValueInput/TokenBalanceAccessory.js (hunk 2 above)
Comment (mdjastrzebski) [reply src/components/Inputs/TokenValueInput/TokenBalanceAccessory.js:75]: I guess that leaving option to ignore the control when there is no balance gives you more flexibility, you can go for zero balance by apply zero default value when getting balance from the store. This is what all current use cases do.

File: src/screens/WalletMigrationArchanova/SetAmount/WalletMigrationArchanovaSetAmount.js [hunk 6]
```diff
@@ -18,94 +18,104 @@
...
+  const [value, setValue] = React.useState<?BigNumber>(initialValue);
 
   // Force active account to be archanova
   React.useEffect(() => {
     dispatch(switchToArchanovaAccountIfNeededAction());
   }, [dispatch]);
 
+  React.useEffect(() => {
+    InteractionManager.runAfterInteractions(() => inputRef.current?.focus());
+  }, []);
```
Comment (poocart) [inline src/screens/WalletMigrationArchanova/SetAmount/WalletMigrationArchanovaSetAmount.js:71]:
```suggestion
  React.useEffect(() => {
    const handler = InteractionManager.runAfterInteractions(() => inputRef.current?.focus());
    return () => { InteractionManager.clearInteractionHandle(handler); };
  }, []);
```

File: src/selectors/balances.js [hunk 7]
```diff
@@ -82,3 +89,10 @@ export const paymentNetworkTotalBalanceSelector: (RootReducerState) => BigNumber
   },
 );
 
+export const useWalletAssetBalance = (chain: ?Chain, assetAddress: ?string): BigNumber => {
+  const accountBalances = useRootSelector(accountAssetsBalancesSelector);
+  if (!chain || !assetAddress) return BigNumber(0);
+
+  const balance = valueForAddress(accountBalances[chain]?.wallet, assetAddress)?.balance;
+  return BigNumber(balance ?? 0);
```
Comment (poocart) [inline src/selectors/balances.js:97]: I think this can be combined with `getWalletBalanceForAsset`.

File: src/utils/rates.js [hunk 8]
```diff
@@ -52,3 +52,17 @@ export const getAssetValueInFiat = (
 
   return wrapBigNumber(assetValue)?.times(rate).toNumber();
 };
+
+export const getAssetValueFromFiat = (
+  fiatValue: ?BigNumber | string | number,
```
Comment (poocart) [inline src/utils/rates.js:57]: Does it has something to do with BigNumber not being present or we can just replace this with `Value`?

**Review by poocart — APPROVED** [review]

File: src/screens/WalletMigrationArchanova/SetAmount/WalletMigrationArchanovaSetAmount.js (hunk 6 above)
Comment (mdjastrzebski) [reply src/screens/WalletMigrationArchanova/SetAmount/WalletMigrationArchanovaSetAmount.js]: Fixed in a slightly different manner, as `clearInteractionHandle` is for `createInteractionHandle`.

File: src/utils/rates.js (hunk 8 above)
Comment (mdjastrzebski) [reply src/utils/rates.js:57]: It's the same as `?Value` or `BigNumber | string | number | null | void`

File: src/components/Icons/TokenIcon.js (hunk 1 above)
Comment (mdjastrzebski) [reply src/components/Icons/TokenIcon.js]: Replaced

File: src/screens/Exchange/Exchange/Exchange.js (hunk 5 above)
Comment (mdjastrzebski) [reply src/screens/Exchange/Exchange/Exchange.js]: Changed

File: src/utils/rates.js (hunk 8 above)
Comment (mdjastrzebski) [reply src/utils/rates.js]: Changed

File: src/selectors/balances.js (hunk 7 above)
Comment (mdjastrzebski) [reply src/selectors/balances.js]: Changed

**Pull request merged**